E2E tests: docs update (#21600)
Automattic/jetpack: tools/e2e-commons/README.md

touch config/playwright.config.js
echo "module.exports = require( 'jetpack-e2e-commons/config/playwright.config.default' );" >> config/playwright.config.js
```

## Run the tests

Now you can run the test in2 different ways: against a local site, in a Docker environment, or against a remote site, that you preconfigured.

## 1. Use the local Docker environment

### 1.1. Build Jetpack

```shell
pnpm jetpack build plugins/jetpack
```

### 1.2. Build your plugin

Assuming you're building tests for a standalone plugin, don't forget to also build that.

### 1.3. Start the local environment

Sensitive information like credentials and other secrets is stored in an encrypted config file. This file needs to be decrypted before starting the environment.
If you're an a12n you can find the key in the secret store and set it in the `CONFIG_KEY` env var, as shown below. 

To bypass the offline mode you will need your site to have a publicly accessible url that will proxy all requests to your locally running WordPress installation.
We use `localtunnel` library to expose `localhost:8889` via a public url.

```shell
## Decrypt default config file

> CONFIG_KEY=secret_key openssl enc -md sha1 -aes-256-cbc -d -pass env:CONFIG_KEY -in ./node_modules/jetpack-e2e-commons/config/encrypted.enc -out config/local-test.js

Author: Yeah, ideally, but at this step in the guide it isn't, and it should be up to the tests developer to add it as a script or not. 
I plan to add a wrapper script over all these decryption/encryption commands, to make them simpler. That will be in another PR, I wanted to put this doc out there first.

## Start and the Docker environment and configure the WordPress installation
pnpm e2e-env start

## Create a tunnel
pnpm tunnel on 
```

The tunnel url will be stored in a file in the config folder of your tests, so that it can be read by the tests and then reused by the tunnel script. See config files for details.

### 1.4. Run the tests

```shell
NODE_CONFIG_DIR='./config' pnpm jest
```

## 2. Use a remote preconfigured site

### 2.1. Add the test site details in the config file

> Edit the encrypted config file and add an entry in the `testSites` object with the details of your test site

Reviewer: ```suggestion
Edit the decrypted config file to add an entry in the `testSites` object with the details of your test site
```

```js
mySite: {
    url: 'site-url', 
    username: 'username',
    password: 'password',
    dotComAccount: ['username', 'password']
}
```

### 2.2. Run the tests

Set the `TEST_SITE` environment variable with the name of the previously defined configuration object.

```shell
TEST_SITE=mySite NODE_CONFIG_DIR='./config' pnpm jest
```

# Functionality plugins

Tests rely on functionality plugins that provide some additional functionality, provide shortcuts, etc.

### e2e-plan-data-interceptor.php

> The purpose of this plugin is to provide a way to `mock` Jetpack plan, for cases when we test functionality that does not directly use paid services. Great example of this purpose is a paid Gutenberg blocks.

Reviewer: ```suggestion
The purpose of this plugin is to provide a way to `mock` a Jetpack plan, for cases when we test functionality that does not directly use paid services. A great example of this purpose is testing paid Gutenberg blocks.
```

# Test architecture

## Pages

The tests are using the `PageObject` pattern, which is a way to separate test logic from implementation. Page objects are basically abstractions around specific pages and page components.
There are two base classes that should be extended by page objects: [`WpPage`](./pages/wp-page.js) and [`PageActions`](./pages/page-actions.js) class.

`WpPage` implements common page methods, like `init` - static method that initializes a page object and checks the displayed page is the expected one, and `visit` - method that navigates to a page URL and then performs all the `init` checks.

`WpPage` extends `PageActions`.
`WpPage` should be extended by all page objects that represent full pages. Rule of thumb: if it has a URL it should extend WpPage. Otherwise, it's probably representing a page component (like a block) and should directly extend `PageActions`.

> Since most of the Playwright functionality is `async`, and JavaScript constructors are not - we should initialize pages with `init()` static method: `await BlockEditorPage.init( page )` to make sure we would wait for `expectedSelectors` checks.

Reviewer: ```suggestion
Since most of the Playwright functionality is `async` - and JavaScript constructors are not - we should initialize pages with the `init()` static method: `await BlockEditorPage.init( page )` to make sure we would wait for `expectedSelectors` checks.
```
or
```suggestion
Since most of the Playwright functionality is `async` and JavaScript constructors are not, we should initialize pages with the `init()` static method: `await BlockEditorPage.init( page )` to make sure we would wait for `expectedSelectors` checks.
```
